### PR TITLE
Add layout default value support for blocks

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description

Based on this comment: https://github.com/WordPress/gutenberg/pull/34189#issuecomment-902528390

This PR adds support for layout default value for a block. This will be later used for blocks like `Social Icons` ([see first explorations WIP](https://github.com/WordPress/gutenberg/pull/33987)) where currently 'hack' the `flex` layout styles instead of actually making it a declarative layout. 

A couple more flags were added:
1. `allowSwitching` - it will show the `layout switcher`. This will probably be augmented in future explorations with a possible `supportedLayouts` in a block.
2. `allowInheriting` - shows the `inherit` toggle
3.  `allowEditing` - hides completely all the `layout`

Example for `block.json` in `supports`:
```
"__experimentalLayout": {
	"allowSwitching": false, // this is the default
	"allowInheriting": true, // this is the default
        "allowEditing": true, // this is the default
	"default": {
		"type": "flex",
		"column-gap": "normal"
	}
}
```

### CODE REVIEW

File: lib/block-supports/layout.php [hunk 1]
```diff
@@ -69,7 +69,6 @@ function gutenberg_get_layout_style( $selector, $layout ) {
 		$style .= 'display: flex;';
 		$style .= 'gap: var( --wp--style--block-gap, 0.5em );';
 		$style .= 'flex-wrap: wrap;';
-		$style .= 'align-items: center;';
```
Comment (youknowriad) [inline lib/block-supports/layout.php]: This is unrelated right?

File: packages/block-editor/src/hooks/layout.js [hunk 2]
```diff
@@ -53,8 +44,13 @@ function LayoutPanel( { setAttributes, attributes, name: blockName } ) {
 		return null;
 	}
 
-	const allowLayoutSwitching = canBlockSwitchLayout( blockName );
-	const { inherit = false, type = 'default' } = layout;
+	const {
+		allowSwitching: canBlockSwitchLayout,
+		default: defaultBlockLayout,
+	} = getBlockSupport( blockName, layoutBlockSupportKey ) || {};
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/layout.js:52]: I think we should have a flag to disable all layout UI: Basically you should be able to choose a default layout but do not allow changes by the user.

File: packages/block-editor/src/hooks/layout.js [hunk 3]
```diff
@@ -166,21 +162,20 @@ export const withInspectorControls = createHigherOrderComponent(
...
+		const defaultThemeLayout = useSetting( 'layout' ) || {};
 		const element = useContext( BlockList.__unstableElementContext );
-		const shouldRenderLayoutStyles = supportLayout || hasInnerBlocks;
 		const { layout = {} } = attributes;
-		const usedLayout = !! layout && layout.inherit ? defaultLayout : layout;
+		const { default: defaultBlockLayout } =
+			getBlockSupport( name, layoutBlockSupportKey ) || {};
+		const usedLayout = layout?.inherit
+			? defaultThemeLayout
+			: defaultBlockLayout || {};
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/layout.js]: I think there's a bug here, we should first default to `layout` attribute before defaulting to the "defaultBlockLayout"

File: lib/block-supports/layout.php (hunk 1 above)
Comment (ntsekouras) [reply lib/block-supports/layout.php]: Yes, but it's not needed. This style is handled with block attributes in some blocks now and eventually will be handled by the `layout.flex` controls. For example if we added the support for `Buttons` it would override the style provided by the block.

Comment (youknowriad) [reply lib/block-supports/layout.php]: I think we should remove it once we introduce the possibility to tweak otherwise we'll be breaking existing flex blocks as otherwise, their content won't be centered. (which we don't really have but you could try by allowing switching for group)

Comment (ntsekouras) [reply lib/block-supports/layout.php]: I reverted for now and will revisit when the time comes.

File: packages/block-editor/src/hooks/layout.js (hunk 2 above)
Comment (ntsekouras) [reply packages/block-editor/src/hooks/layout.js:52]: I add an `allowEditing = true` flag and one more `allowInheriting` for displaying the respective control. Seems like many flags are needed.. Do you have any thoughts if we can simplify these?

Comment (youknowriad) [reply packages/block-editor/src/hooks/layout.js:52]: I'm not sure yet but yes, it's something we need to think about while it's still experimental.

**Review by youknowriad — COMMENTED** [review]
This is looking good. Should we add 

```
__experimentalLayout: {
     default: { type: 'flow' },
     allowEditing: false
}
```

to all the Core blocks like "Cover", "Media & Text", "Post Template"... (all the container blocks that are vertical/flow) in order to support the "block gap" theme.json config properly?

**Comment by youknowriad** [comment]
mmm tried testing a bit here while it works well for post-template, it seems to break the design for cover an media & text.
The reason for that seems to be that the layout styles are applied the "wrapper of the block" instead of the "wrapper of the inner blocks". Meaning gap styles don't work properly for Cover and Media & Text.

So I guess these two blocks will have to wait for another PR (potentially trying to find a way to apply the layout styles to the right wrapper), for now we should remove the `__experimentalLayout` from these two blocks.

**Review by youknowriad — APPROVED** [review]
